datafusion/common/src/error.rs

```diff
@@ -758,6 +758,116 @@ macro_rules! unwrap_or_internal_err {
     };
 }
 
+/// Assert a condition, returning `DataFusionError::Internal` on failure.
+///
+/// # Examples
+///
+/// ```text
+/// assert_or_internal_err!(predicate);
+/// assert_or_internal_err!(predicate, "human readable message");
+/// assert_or_internal_err!(predicate, format!("details: {}", value));
+/// ```
+#[macro_export]
+macro_rules! assert_or_internal_err {
+    ($cond:expr) => {
+        if !$cond {
+            return Err(DataFusionError::Internal(format!(
+                "Assertion failed: {}",
+                stringify!($cond)
+            )));
+        }
+    };
+    ($cond:expr, $msg:expr) => {
+        if !$cond {
+            return Err(DataFusionError::Internal(format!(
+                "Assertion failed: {}: {}",
+                stringify!($cond),
+                $msg
```

Comment:
Keep all three macros support the same format argument syntax

nit:
```suggestion
    ($cond:expr, $($arg:tt)+) => {
        if !$cond {
            return Err(DataFusionError::Internal(format!(
                "Assertion failed: {}: {}",
                stringify!($cond),
                format!($($arg)+)
```